### PR TITLE
Clipped and Softer Max

### PR DESCRIPTION
**Pull request by photomz**
Adds clipped softmax and softermax.

- Unit PyTests with edge cases and snapshots
- Learnable parameters in `modelling_softerllama`.

Other changes

- Add `pyproject.toml` and `requirements.txt` -- just housekeeping
- Move `tests` to `src/tests` so imports correctly

@Clyde013 Please reply my comments before merging.

### CODE REVIEW

File: src/functional.py [hunk 1]
```diff
@@ -3,36 +3,56 @@
...
-        original_denominator, torch.multiply(torch.exp(shifted_zeros), n_bias)
-    )
-    output = torch.divide(numerator, denominator)
-    return output if dtype is None else output.type(dtype=dtype)
+    scores = numerator / (denominator + shifted_bias)
+    return scores
+
+
+# Ref: https://github.com/Qualcomm-AI-research/outlier-free-transformers/blob/1e59744972da808a90d9996027bfc274689594b8/transformers_language/models/softmax.py#L8
+def clip_logits(score: torch.Tensor, gamma: torch.Tensor = zero, eta: torch.Tensor = one) -> torch.Tensor:
```
Comment (photomz) [inline src/functional.py:38]: Prefer clipping without softmax, allows us to stack many funcs like: `clip(softermax(QK'))` if we ever want to train that combo.

Old clip+softmax combined func is still below.

File: src/softerllama/modeling_softerllama.py
Comment (photomz) [inline src/softerllama/modeling_softerllama.py:1]: Sorry for the big diff, my linter reformatted the file

File: src/softerllama/modeling_softerllama.py [hunk 2]
```diff
@@ -51,8 +51,9 @@ def __init__(self, config: SofterLlamaConfig, layer_idx: Optional[int] = None):
                 "when creating this class."
             )
 
-        # softermax n_bias parameter
-        self.n_bias = config.n_bias
+        # MZ: softermax n_bias, clipped softmax n_clip parameters
+        self.n_bias = nn.Parameter(torch.tensor(config.n_bias), requires_grad=config.learn_softmax)
```
Comment (photomz) [inline src/softerllama/modeling_softerllama.py:55]: Will be fixed constants when `require_grad=False` --- cause a Parameter will be no different than a Tensor then.

File: src/softerllama/modeling_softerllama.py [hunk 3]
```diff
@@ -224,12 +194,9 @@ def forward(
 
         # upcast attention to fp32
```
Comment (photomz) [inline src/softerllama/modeling_softerllama.py:195]: Why did you upcast to `fp32`, only to downcast back to `query_states.dtype`? Seems redundant

File: src/softerllama/modeling_softerllama.py [hunk 4]
```diff
@@ -224,12 +194,9 @@ def forward(
...
         # softermax overriding original attention softmax
-        attn_weights = softermax(
-            attn_weights, self.n_bias, dim=-1, dtype=torch.float32
-        ).to(query_states.dtype)
-        attn_weights = nn.functional.dropout(
-            attn_weights, p=self.attention_dropout, training=self.training
-        )
+        attn_weights = softermax(attn_weights, self.n_bias, dim=-1).type(dtype=torch.float32).to(query_states.dtype)
+        attn_weights = clip_logits(attn_weights, self.n_clip, dim=-1)
```
Comment (photomz) [inline src/softerllama/modeling_softerllama.py:198]: `clip` stacks on `softermax` -- to turn off clip we just set `n_clip=0`, to turn off softermax we just set `n_bias=0`.

@Clyde013 Is this inefficient? Or do CUDA optimisations / fused kernels negate that at the limit

**Review by photomz — COMMENTED** [review]
Please check through @Clyde013 !

File: src/functional.py [hunk 5]
```diff
@@ -3,36 +3,56 @@
...
 
 
 # Ref: https://github.com/softmax1/EsperBERTo/blob/946f0a9fa0f6e3b2bf755388d8fa20c31f8e2bf5/src/functional.py#L49
-def softermax(
-    input: Tensor,
-    n_bias: float = 0.0,
-    dim: Optional[int] = None,
-    dtype: Optional[_dtype] = None,
-) -> Tensor:
+def softermax(input: torch.Tensor, n_bias: torch.Tensor = zero, dim: Optional[int] = -1) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (Clyde013) [reply src/functional.py:15]: LGTM

File: src/functional.py (hunk 1 above)
Comment (Clyde013) [reply src/functional.py:38]: Makes sense to me. 👍

File: src/softerllama/modeling_softerllama.py (hunk 2 above)
Comment (Clyde013) [reply src/softerllama/modeling_softerllama.py:55]: yeah sounds correct 👍

File: src/softerllama/modeling_softerllama.py (hunk 4 above)
Comment (Clyde013) [reply src/softerllama/modeling_softerllama.py:198]: Realistically the inefficiencies are going to be incredibly minor overhead, and if it does turn out to be a big deal we can wrap the whole thing inside an if-statement. Not a big enough deal for us to care too much about it.

File: src/softerllama/modeling_softerllama.py (hunk 3 above)
Comment (Clyde013) [reply src/softerllama/modeling_softerllama.py:195]: Legacy code inherited from original huggingface llama implementation. Doesn't hurt to keep it there, I'm not sure what impact changing this behaviour could cause, and I'm not sure we want to find out either. Apply the age-old adage: If it works, don't touch it.

**Review by Clyde013 — APPROVED** [review]
LGTM 👍
